source/extensions/filters/http/common/aws/credentials_provider_impl.h

```diff
@@ -0,0 +1,172 @@
+#pragma once
+
+#include <list>
+
+#include "envoy/api/api.h"
+#include "envoy/event/timer.h"
+
+#include "common/common/logger.h"
+#include "common/common/thread.h"
+
+#include "extensions/filters/http/common/aws/credentials_provider.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Common {
+namespace Aws {
+
+/**
+ * Retrieve AWS credentials from the environment variables.
+ *
+ * Adheres to conventions specified in:
+ * https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html
+ */
+class EnvironmentCredentialsProvider : public CredentialsProvider,
+                                       public Logger::Loggable<Logger::Id::aws> {
+public:
+  Credentials getCredentials() override;
+};
+
+class MetadataCredentialsProviderBase : public CredentialsProvider,
+                                        public Logger::Loggable<Logger::Id::aws> {
+public:
+  typedef std::function<absl::optional<std::string>(const std::string& host,
+                                                    const std::string& path,
+                                                    const absl::optional<std::string>& auth_token)>
+      MetadataFetcher;
+
+  MetadataCredentialsProviderBase(Api::Api& api, const MetadataFetcher& metadata_fetcher)
+      : api_(api), metadata_fetcher_(metadata_fetcher) {}
+
+  Credentials getCredentials() override {
+    refreshIfNeeded();
+    return cached_credentials_;
+  }
+
+protected:
+  Api::Api& api_;
+  MetadataFetcher metadata_fetcher_;
+  SystemTime last_updated_;
+  Credentials cached_credentials_;
+  Thread::MutexBasicLockable lock_;
+
+  void refreshIfNeeded();
+
+  virtual bool needsRefresh() PURE;
+  virtual void refresh() PURE;
+};
+
+/**
+ * Retrieve AWS credentials from the instance metadata.
+ *
+ * https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html#instance-metadata-security-credentials
+ */
+class InstanceProfileCredentialsProvider : public MetadataCredentialsProviderBase {
+public:
+  InstanceProfileCredentialsProvider(Api::Api& api, const MetadataFetcher& metadata_fetcher)
+      : MetadataCredentialsProviderBase(api, metadata_fetcher) {}
+
+private:
+  bool needsRefresh() override;
+  void refresh() override;
+};
+
+/**
+ * Retrieve AWS credentials from the task metadata.
+ *
+ * https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-iam-roles.html#enable_task_iam_roles
+ */
+class TaskRoleCredentialsProvider : public MetadataCredentialsProviderBase {
+public:
+  TaskRoleCredentialsProvider(
+      Api::Api& api, const MetadataFetcher& metadata_fetcher, const std::string& credential_uri,
+      const absl::optional<std::string>& authorization_token = absl::optional<std::string>())
```

Comment:
Seems like we can just use empty string here rather than juggling with optional?

Reply:
Makes sens. I replaced `absl::optional` with `std::string`